Add a --board parameter to sprint list

andreadecorte opened the pull request:
Add an optional `--board` parameter that takes in input a board id that overrides the parameter specified in the config file. This avoids the need to edit the config file in case a change of board is needed.

Also implement dynamic retrieval of the board name in case the board parameter is used.

Parameter is added only to `sprint list` command as it is the only one requiring a board at the moment.

Fixes #847

andreadecorte commented on internal/cmd/sprint/list/list.go:
```diff
@@ -291,6 +294,32 @@ func setFlags(cmd *cobra.Command) {
+	board, err := client.BoardByID(boardID)
+	if err != nil {
+		return viper.GetString("board.name")
+	}
```
add warn and fallback value

nicomen commented:
Sure thanks. So my issue is that I haven't been able to find anything when for instance using `jira sprint list` – until I found this (and another related bug). I then realised that the config had a board id. I don't remmeber adding that myself. I was able to change the id and finally find the sprints. But I don't get how the original id ended up in that config file.

ankitpokhrel commented on internal/cmd/sprint/list/list.go:
```diff
@@ -277,6 +280,7 @@ func getIssueQuery(project string, flags query.FlagParser, showAll bool) (string
 }
 
 func setFlags(cmd *cobra.Command) {
+	cmd.Flags().String("board", "", "Board ID to use (overrides board.id from config)")
```
Why do we have this flag as `String`, isn't board id always supposed to be integer?

andreadecorte replied:
good catch

andreadecorte commented on pkg/jira/board_test.go:
```diff
@@ -13,48 +13,60 @@ import (
 
+	var unexpectedResponseErr *ErrUnexpectedResponse
+
 	_, err := client.Boards("BAD", "scrum")
-	assert.Error(t, &ErrUnexpectedResponse{}, err)
+	assert.ErrorAs(t, err, &unexpectedResponseErr)
 
 	_, err = client.BoardSearch("BAD", "scrum")
-	assert.Error(t, &ErrUnexpectedResponse{}, err)
+	assert.ErrorAs(t, err, &unexpectedResponseErr)
```
declaring different instances doesn't provide much value in this case

andreadecorte commented:
> Hi @andreadecorte, I think we can slightly refactor the implementation. Below is the diff to help address my PR comments. The logic remains the same, so it shouldn't cause any breaks. Feel free to take a look and see if you’re onboard!
> 
> ```diff
> --- a/internal/cmd/sprint/list/list.go
> +++ b/internal/cmd/sprint/list/list.go
> @@ -72,7 +72,6 @@ func SetFlags(cmd *cobra.Command) {
>  func sprintList(cmd *cobra.Command, args []string) {
>         server := viper.GetString("server")
>         project := viper.GetString("project.key")
> -       boardFlagChanged := cmd.Flags().Changed("board")
>         boardID := getBoardID(cmd)
> 
>         debug, err := cmd.Flags().GetBool("debug")
> @@ -84,7 +83,7 @@ func sprintList(cmd *cobra.Command, args []string) {
>         cmdutil.ExitIfError(err)
> 
>         if len(args) == 0 {
> -               sprintExplorerView(sprintQuery, cmd.Flags(), boardID, boardFlagChanged, project, server, client)
> +               sprintExplorerView(sprintQuery, cmd.Flags(), boardID, project, server, client)
>         } else {
>                 sprintID, err := strconv.Atoi(args[0])
>                 cmdutil.ExitIfError(err)
> @@ -191,7 +190,7 @@ func singleSprintView(sprintQuery *query.Sprint, flags query.FlagParser, boardID
>         cmdutil.ExitIfError(v.Render())
>  }
> 
> -func sprintExplorerView(sprintQuery *query.Sprint, flags query.FlagParser, boardID int, boardFlagChanged bool, project, server string, client *jira.Client) {
> +func sprintExplorerView(sprintQuery *query.Sprint, flags query.FlagParser, boardID int, project, server string, client *jira.Client) {
>         sprints := func() []*jira.Sprint {
>                 s := cmdutil.Info("Fetching sprints...")
>                 defer s.Stop()
> @@ -204,7 +203,7 @@ func sprintExplorerView(sprintQuery *query.Sprint, flags query.FlagParser, board
>                 return
>         }
> 
> -       boardName := getBoardName(client, boardID, boardFlagChanged)
> +       boardName := getBoardName(client, boardID)
> 
>         if sprintQuery.Params().Current || sprintQuery.Params().Prev || sprintQuery.Params().Next {
>                 sprint := sprints[0]
> @@ -280,7 +279,7 @@ func getIssueQuery(project string, flags query.FlagParser, showAll bool) (string
>  }
> 
>  func setFlags(cmd *cobra.Command) {
> -       cmd.Flags().String("board", "", "Board ID to use (overrides board.id from config)")
> +       cmd.Flags().Int("board", 0, "Board ID to use (overrides board.id from config)")
>         cmd.Flags().String("state", "", "Filter sprint by its state (comma separated).\n"+
>                 "Valid values are future, active and closed.\n"+
>                 `Defaults to "active,closed"`)
> @@ -296,20 +295,18 @@ func setFlags(cmd *cobra.Command) {
>  }
> 
>  func getBoardID(cmd *cobra.Command) int {
> -       boardFlag, err := cmd.Flags().GetString("board")
> -       cmdutil.ExitIfError(err)
> -
> -       if boardFlag != "" {
> -               boardID, err := strconv.Atoi(boardFlag)
> +       if cmd.Flags().Changed("board") {
> +               boardID, err := cmd.Flags().GetInt("board")
>                 cmdutil.ExitIfError(err)
> +
>                 return boardID
>         }
> -
>         return viper.GetInt("board.id")
>  }
> ```

makes sense, refactored following this
